### ISSUE TITLE
Add capability for sending HTTP metadata (trailer fields) in mid-stream

### ISSUE THREAD

**Issue by royfielding** [quicwg/base-drafts#3265]
For a very long time (since early 1995) there have been various attempts to send metadata in parallel with the request or response bodies. The use cases have ranged from group chat to chained digital signatures to performance metrics to long download checkpoints to a heck of a lot of other things that I pretend to no longer remember for sanity purposes.

Fixing that in HTTP/1.1 was only possible within chunked extensions. For a variety of reasons that had nothing to do with the use cases, chunked extensions didn't work well over intermediaries and the chicken never hatched. Nevertheless, they did work over HTTP when the applications had control over both sides of a connection. I vaguely recall such metadata being present in HTTPng as well.

For some bizarre reason, I thought that sending metadata in midstream was present in h2 and all was fine and good for keeping it through in QUIC. I mean, for crying out loud, we don't have any of the limitations here that we had to deal with in HTTP/1.1. That's why I didn't raise this issue earlier. Apparently, I was wrong.

I would like for HTTP/3 to have a mechanism for communicating metadata (like trailers) in mid-stream, both for requests (e.g., priority) and responses (e.g., chained sigs).

I believe that this can be done with an additional frame type that carries trailer data, just like HEADERS, but that doesn't imply the end of DATA. However, I don't really have an opinion on what is the right way to do this in HTTP/3, only that it is worthwhile and necessary to do before a last call. I do not believe, from extensive experience, that this feature can be added later. I'd appreciate if people more familiar with QUIC implementation would choose the best specific syntax.

It is not a semantics change for HTTP because the fields are just optional trailers that arrive early. A recipient that is not prepared to handle these fields, for whatever reason, can discard them in the same way as trailers.

**Comment by mnot** [comment]
Parking until discussion in HTTPWG concludes.

**Comment by larseggert** [comment]
I'd like to see if we can "unpark" this issue and move it forward in some way. Any suggestions?

**Comment by MikeBishop** [comment]
I'm of the opinion that this doesn't belong in HTTP/3 if it hasn't been defined by the HTTP WG before we ship (and it won't be).  If/when this takes shape, it will probably be as an extension to HTTP/2 and HTTP/3, and that's okay.

**Comment by ianswett** [comment]
+1 that this should become an extension and doesn't belong in HTTP/3.

**Comment by larseggert** [comment]
So "close with no action"?

**Comment by royfielding** [comment]
Just leave it parked for now. It is still an issue for QUIC.

**Comment by larseggert** [comment]
We're trying to clear out the "parked" queue, because folks seem to think "parked" means "needs to be resolved for v1."

**Comment by royfielding** [comment]
The discussion in HTTP was entirely positive and there is already a non-IETF spec for it as an extension to h2, but that was in Singapore and nobody has had time to work on it since. The http-core specs already allow for it. h2 is not under revision. The only bits it requires in HTTP/3 is a frame type and some expectation of what to do with it (i.e., it contains fields, treat them as optional trailers, done).

**Comment by ianswett** [comment]
I don't expect most people to use this, and those that do(yes, that includes us) would likely want support in H2 and H3, so I'd prefer an extension for both H2 and H3.

**Comment by larseggert** [comment]
I'm going to mark this as v2 for now. We can unmark it if there is HTTP progress that would allow us to target QUICv1.

**Comment by royfielding** [comment]
That's fine with me. Thanks.

**Comment by LPardue** [comment]
Since this is a design issue, we'll run this through the Late-stage process and include it in the next consensus call. 

The proposal is to:

> mark this as v2 for now. We can unmark it if the HTTP work progresses faster than we do, and we decide to incorporate it in some way.
